### PR TITLE
Feature: BPF_PERF_ARRAY implementation

### PR DESCRIPTION
**Pull request by Skrinikov**
Adds the ability to use `BPF_PERF_ARRAY(name, NUM_CPU)` macros in BPF code.

Example rust code:
```
let event_array = PerfEventArray::new()
   .table("cycles")
   .event(Event::Hardware(HardwareEvent::RefCpuCycles))
   .open(&mut bpf);
```

### CODE REVIEW

File: src/core/perf_event_array/v0_4_0.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            unsafe { bpf_close_perf_event_fd(fd) };
+            return Err(());
+        }
+        self.cpu_fd.insert(cpu, fd);
+
+        Ok(())
+    }
+
+    fn update(&self, cpu: usize, fd: i32) -> bool {
+        (unsafe { bpf_update_elem(self.table_fd, cpu as *mut c_void, fd as *mut c_void, 0) }) >= 0
```
Comment (Skrinikov) [inline src/core/perf_event_array/v0_4_0.rs:90]: Not sure if the cast on this line is valid.

Comment (Skrinikov) [reply src/core/perf_event_array/v0_4_0.rs:86]: Update, this does not seem to work

Comment (brayniac) [reply src/core/perf_event_array/v0_4_0.rs:86]: Maybe something like we do here? https://github.com/rust-bpf/rust-bcc/blob/34786edb555ff611675bc3d179fdc63d575ceaa6/src/table.rs#L60

File: src/perf_event/perf_event_array.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /// on the system to provide system-wide coverage.
+    pub fn cpu(mut self, cpu: Option<usize>) -> Self {
+        self.cpu = cpu;
+        self
+    }
+
+    /// Consumes the perf event and opens it. May return an error if there is a
+    /// incomplete or invalid configuration or other error while loading or
+    /// opening the event.
+    pub fn open(self, bpf: &mut BPF) -> Result<(), BccError> {
```
Comment (brayniac) [inline src/perf_event/perf_event_array.rs:47]: I'd like to keep the terminology consistent across all of these builder-esque types so it's less burden for users. Can we rename to `attach` to match?

File: src/perf_event/perf_event_array.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            });
+        }
+
+        let table = self.table.unwrap();
+        let event = self.event.unwrap();
+
+        let table_fd = bpf.table_fd(&table);
+
+        let ev_type = match event {
+            Event::Hardware(_) => EventType::Hardware,
```
Comment (brayniac) [inline src/perf_event/perf_event_array.rs:66]: We might want to add a method for the `Event` enum that returns the `EventType` so we don't need to match in multiple places.

File: src/perf_event/perf_event_array.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            Event::Software(_) => EventType::Software,
+            Event::HardwareCache(_, _, _) => EventType::HardwareCache,
+        } as u32;
+
+        let ev_config = match event {
+            Event::Hardware(hw_event) => hw_event as u32,
```
Comment (brayniac) [inline src/perf_event/perf_event_array.rs:72]: Similar to the above, we could probably implement a `.ev_config()` function to keep this logic in a single spot.

File: src/core/perf_event_array/v0_4_0.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            return Err("perf event is already open for cpu".to_string());
+        }
+
+        let fd =
+            unsafe { bpf_open_perf_event(self.ev_type, self.ev_config.into(), -1, cpu as i32) };
+
+        if fd < 0 {
+            return Err("failed to open perf on cpu".to_string());
+        }
+        let mut cpu_bytes: [u8; 8] = unsafe { std::mem::transmute(cpu.to_le()) };
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:74]: Wondering about the potential for endianness issues here. Thoughts?

File: src/perf_event/perf_event_array.rs (hunk 3 above)
Comment (brayniac) [reply src/perf_event/perf_event_array.rs:66]: Maybe this should just return a u32 - not sure what else could really be done for the corresponding `.ev_config()` method.

File: src/core/perf_event_array/v0_4_0.rs (hunk 5 above)
Comment (Skrinikov) [reply src/core/perf_event_array/v0_4_0.rs:74]: I was wondering the same thing. usize did not have `to_ne` so I added this as a placeholder. I just realized that there is a `to_ne_bytes()` function so I will try to see if I can change it to that.

**Comment by brayniac** [comment]
@Skrinikov - anything more on this before it's ready to move out of draft?

File: src/core/perf_event_array/v0_4_0.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                unsafe { bpf_close_perf_event_fd(*v) };
+            }
+
+            false
+        });
+    }
+
+    pub fn close_on_cpu(&mut self, cpu: usize) {
+        let fd = self.cpu_fd.remove(&cpu);
+        if fd.is_some() {
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:41]: possible to combine this into `if let Some(fd) = self.cpu_fd.remove(&cpu) {` ?

File: src/core/perf_event_array/v0_4_0.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            unsafe { bpf_close_perf_event_fd(fd.unwrap()) };
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:42]: you could get rid of the unwrap here if the above works

File: src/core/perf_event_array/v0_4_0.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                    return Err(e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn open_on_cpu(&mut self, cpu: usize) -> Result<(), String> {
+        if self.cpu_fd.get(&cpu).is_some() {
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:64]: `.contains_key()` insead ?

File: examples/cpi.rs [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    );
+
+    let mut bpf = BPF::new(&code)?;
+    PerfEventArray::new()
+        .table("instr_perf")
+        .event(Event::Hardware(HardwareEvent::Instructions))
+        .attach(&mut bpf)?;
+    PerfEventArray::new()
+        .table("cycle_perf")
+        .event(Event::Hardware(HardwareEvent::RefCpuCycles))
```
Comment (brayniac) [inline examples/cpi.rs:50]: Can we add fall-back to `HardwareEvent::CpuCycles`, possibly with a warning to the user to confirm the behavior for their microarchitecture wrt frequency scaling?

File: src/core/perf_event_array/v0_4_0.rs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    pub fn close_on_cpu(&mut self, cpu: usize) {
+        if let Some(fd) = self.cpu_fd.remove(&cpu) {
+            unsafe { bpf_close_perf_event_fd(fd) };
+        }
+    }
+
+    pub fn open_all_cpu(&mut self) -> Result<(), String> {
+        let cpus = cpuonline::get();
+
+        if let Ok(cpus) = cpus {
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:49]: Any reason to not combine line 47 & 49 here?

File: src/core/perf_event_array/v0_4_0.rs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                let result = self.open_on_cpu(cpu);
+
+                if let Err(e) = result {
+                    self.close_all_cpu();
+                    return Err(e);
+                }
+            }
+        }
+
+        Ok(())
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:60]: Is returning Ok if cpuonline::get() fails the correct behavior? Should we maybe raise an error instead?

File: src/core/perf_event_array/v0_4_0.rs [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            return Err("failed to open perf on cpu".to_string());
+        }
+        let mut cpu_bytes: [u8; 8] = cpu.to_ne_bytes();
+        let mut fd_bytes: [u8; 4] = fd.to_ne_bytes();
+
+        let errno = self.update(&mut cpu_bytes, &mut fd_bytes);
+        if errno < 0 {
+            unsafe { bpf_close_perf_event_fd(fd) };
+            return Err(format!(
+                "Unable to open perf event on CPU `{}`, errno {}",
```
Comment (brayniac) [inline src/core/perf_event_array/v0_4_0.rs:81]: for consistency, lowercase "unable" and "cpu"?

File: src/table.rs [hunk 13]
```diff
@@ -156,9 +156,11 @@ impl EntryIter {
         };
         self.current = Some(entry);
         unsafe {
-            let (k, _) = self.entry_ptrs().unwrap();
+            let (k, l) = self.entry_ptrs().unwrap();
```
Comment (brayniac) [inline src/table.rs:159]: nit: `v` for value?

**Pull request merged**